src/main/java/com/yahoo/bullet/bql/query/TypeChecker.java

```diff
@@ -59,6 +81,27 @@
         return Optional.empty();
     }
 
+    static Optional<List<BulletError>> validateBetweenType(BetweenPredicateNode node, Expression value, Expression lower, Expression upper) {
+        Type valueType = value.getType();
+        Type lowerType = lower.getType();
+        Type upperType = upper.getType();
+        if (Type.isUnknown(valueType) || Type.isUnknown(lowerType) || Type.isUnknown(upperType)) {
+            return unknownError();
+        }
+        List<BulletError> errors = new ArrayList<>();
+        if (!Type.isNumeric(valueType)) {
+            errors.add(makeErrorOnly(node, QueryError.BETWEEN_VALUE_NOT_NUMERIC, node, valueType));
+        }
+        if (!Type.isNumeric(lowerType)) {
+            errors.add(makeErrorOnly(node, QueryError.BETWEEN_START_NOT_NUMERIC, node, lowerType));
+        }
+        if (!Type.isNumeric(upperType)) {
+            errors.add(makeErrorOnly(node, QueryError.BETWEEN_END_NOT_NUMERIC, node, upperType));
+        }
+        return !errors.isEmpty() ? Optional.of(errors) : Optional.empty();
```

Comment:
We can look into pre-evaluating expressions when we build queries. We don't have BETWEEN for strings rn but should be do-able. Dates I'm not so sure because we don't have a date datatype like sql does.